File: apps/web/lib/textSafety.ts
/**
 * Sensitive text detection and redaction for public metadata, previews, and analytics.
 * 
 * Frontend equivalent of the backend utils/text_safety.py module.
 * Used for:
 * - OG metadata generation in Next.js generateMetadata
 * - Public run title previews
 * - Analytics event payloads (prevent prompt leakage)
 */

import {
  createDetectionPatterns,
  createRedactionPatterns,
  type CompiledSafetyPattern,
} from "./generatedPatterns";


function isLuhnValid(candidate: string): boolean {
  const digits = candidate.replace(/[ -]/g, "");
  if (!/^\d{13,19}$/.test(digits)) return false;

  let checksum = 0;
  const parity = digits.length % 2;
  for (let index = 0; index < digits.length; index += 1) {
    let value = Number(digits[index]);
    if (index % 2 === parity) {
      value *= 2;
      if (value > 9) value -= 9;
    }
    checksum += value;
  }
  return checksum % 10 === 0;
}


function isValidMatch(pattern: CompiledSafetyPattern["pattern"], value: string): boolean {
  return pattern.validator !== "luhn" || isLuhnValid(value);
}


function hasValidMatch({ pattern, regex }: CompiledSafetyPattern, text: string): boolean {
  let remaining = text;
  let match = regex.exec(remaining);
  while (match) {
    if (isValidMatch(pattern, match[0])) return true;
    const consumed = match.index + Math.max(match[0].length, 1);
    remaining = remaining.slice(consumed);
    match = regex.exec(remaining);
  }
  return false;
}


/**
 * Check if text contains any sensitive patterns.
 * 
 * Returns true if the text likely contains PII, API keys, tokens,
 * or other secrets that should not appear in public metadata.
 */
export function containsSensitivePattern(text: string): boolean {
  if (!text) return false;
  return createDetectionPatterns().some((pattern) => hasValidMatch(pattern, text));
}


export function detectSensitiveCategories(text: string): string[] {
  if (!text) return [];
  const categories = new Set<string>();
  for (const compiled of createDetectionPatterns()) {
    if (hasValidMatch(compiled, text)) categories.add(compiled.pattern.category);
  }
  return Array.from(categories).sort();
}


export function sanitizePublicText(text: string): string {
  let result = text;
  for (const { pattern, regex } of createRedactionPatterns()) {
    result = result.replace(
      regex,
      (match) => isValidMatch(pattern, match) ? pattern.replacement : match,
    );
  }
  return result;
}


/**
 * Create a safe, truncated preview of text for metadata.
 * 
 * 1. Checks for sensitive patterns — falls back to generic text
 * 2. Truncates to maxLength
 * 3. Cleans whitespace
 */
export function truncatePublicPreview(text: string, maxLength = 60): string {
  if (!text) return "Shared Arena Run";

  if (containsSensitivePattern(text)) {
    return "Shared Arena Run";
  }

  const clean = text.trim().replace(/\n/g, " ").replace(/\r/g, "").replace(/\s+/g, " ");

  if (clean.length <= maxLength) return clean;

  return clean.slice(0, maxLength - 3).trimEnd() + "...";
}


/**
 * Generate a safe page title for a debate/run.
 * 
 * For public runs with safe prompts: "Arena Run: {preview} | Consultaion"
 * For public runs with sensitive prompts: "Shared Arena Run | Consultaion"
 * For private runs: "Arena Run | Consultaion" (never expose prompt)
 */
export function safeMetadataTitle(prompt: string, isPublic = true): string {
  if (!isPublic) return "Arena Run | Consultaion";

  const preview = truncatePublicPreview(prompt, 57);
  if (preview === "Shared Arena Run") {
    return "Shared Arena Run | Consultaion";
  }

  return `Arena Run: ${preview} | Consultaion`;
}


/**
 * Generate a safe meta description for a debate/run.
 */
export function safeMetadataDescription(): string {
  return "Compare multiple AI model responses and read the synthesized answer.";
}

File: apps/api/utils/text_safety.py
"""
Sensitive text detection and redaction for public metadata, previews, and analytics.

Builds on the existing PII scrubber (safety.pii) and adds patterns for:
- API keys (OpenAI, Anthropic, Google, etc.)
- JWT tokens and bearer tokens
- Credit card numbers
- High-entropy secrets and passwords
- URLs with embedded tokens

This module is used for:
1. OG metadata title/description generation
2. Public run previews
3. Analytics event payloads (to prevent prompt leakage)
4. Audit log sanitization
"""

from __future__ import annotations

import re

from safety.patterns import (
    contains_sensitive_pattern as _contains_sensitive_pattern,
    redact_sensitive_patterns,
)

# ---------------------------------------------------------------------------
# Detection
# ---------------------------------------------------------------------------

def contains_sensitive_pattern(text: str) -> bool:
    """
    Check if text contains any sensitive patterns.

    Returns True if the text likely contains PII, API keys, tokens,
    or other secrets that should not appear in public metadata.
    """
    return _contains_sensitive_pattern(text)


# ---------------------------------------------------------------------------
# Redaction
# ---------------------------------------------------------------------------

def sanitize_public_text(text: str) -> str:
    """
    Redact sensitive content from text for safe public display.

    Replaces API keys, tokens, PII, secrets with [REDACTED] placeholders.
    Used for metadata generation, public previews, analytics payloads.
    """
    return redact_sensitive_patterns(text)


def truncate_public_preview(text: str, max_length: int = 60) -> str:
    """
    Create a safe, truncated preview of text for metadata.

    1. Sanitizes sensitive content
    2. Truncates to max_length
    3. Falls back to generic text if content appears sensitive

    Returns a safe string for use in OG titles, descriptions, etc.
    """
    if not text:
        return "Shared Arena Run"

    # Check if the original text contains sensitive patterns
    if contains_sensitive_pattern(text):
        return "Shared Arena Run"

    # Clean and truncate
    clean = text.strip().replace("\n", " ").replace("\r", "")
    # Collapse multiple spaces
    clean = re.sub(r"\s+", " ", clean)

    if len(clean) <= max_length:
        return clean

    return clean[:max_length - 3].rstrip() + "..."


def safe_metadata_title(prompt: str, is_public: bool = True) -> str:
    """
    Generate a safe page title for a debate/run.

    For public runs with safe prompts: "Arena Run: {preview} | Consultaion"
    For public runs with sensitive prompts: "Shared Arena Run | Consultaion"
    For private runs: "Arena Run | Consultaion" (never expose prompt)
    """
    if not is_public:
        return "Arena Run | Consultaion"

    preview = truncate_public_preview(prompt, max_length=57)
    if preview == "Shared Arena Run":
        return "Shared Arena Run | Consultaion"

    return f"Arena Run: {preview} | Consultaion"


def safe_metadata_description(prompt: str, is_public: bool = True) -> str:
    """
    Generate a safe meta description for a debate/run.

    For public runs with safe prompts: uses a preview + generic suffix.
    For sensitive/private runs: uses a fully generic description.
    """
    generic = "Compare multiple AI model responses and read the synthesized answer."

    if not is_public:
        return generic

    if contains_sensitive_pattern(prompt):
        return generic

    return generic
